Dip/DipLinuxTests/Sources/main.swift

```diff
@@ -0,0 +1,156 @@
+}
+
+func dispatch_sync(block: TMain) -> UnsafeMutablePointer<Void> {
+  var pid: pthread_t = 0
+  var result: UnsafeMutablePointer<Void> = nil
+  pthread_create(&pid, nil, block, nil)
+  pthread_join(pid, &result)
+  return result
+}
+
+private var resolvedServices = Set<ServiceImp>()
+private var resolvedClients = Array<ClientImp>()
+
+private var lock: pthread_spinlock_t = 0
+pthread_spin_init(&lock, 0)
+
+private var container: DependencyContainer!
+
+class ThreadSafetyTests: XCTestCase {
+  
+  var allTests: [(String, () throws -> Void)] {
+    return [
+      ("testSingletonThreadSafety", testSingletonThreadSafety),
+      ("testFactoryThreadSafety", testFactoryThreadSafety),
+      ("testCircularReferenceThreadSafety", testCircularReferenceThreadSafety)
+    ]
+  }
+  
+  func setUp() {
+    container = DependencyContainer()
+  }
+  
+  func tearDown() {
+  }
+  
+  func testSingletonThreadSafety() {
+    container.register(.Singleton) { ServiceImp() as Service }
+    
+    for _ in 0..<100 {
+      dispatch_async { _ in
+        let service = try! container.resolve() as Service
+        pthread_spin_lock(&lock)
+        resolvedServices.insert(service as! ServiceImp)
+        pthread_spin_unlock(&lock)
+        return nil
+      }
+    }
+    
+    sleep(1)
+    XCTAssertEqual(resolvedServices.count, 1, "Should create only one instance")
+    resolvedServices.removeAll()
+  }
+  
+  
+  func testFactoryThreadSafety() {
+    container.register { ServiceImp() as Service }
+    
+    for _ in 0..<100 {
+      dispatch_async { _ in
+        let service = try! container.resolve() as Service
+        pthread_spin_lock(&lock)
+        resolvedServices.insert(service as! ServiceImp)
+        pthread_spin_unlock(&lock)
+        return nil
+      }
+    }
+    
+    sleep(1)
+    XCTAssertEqual(resolvedServices.count, 100, "All instances should be different")
+    resolvedServices.removeAll()
+  }
+  
+  
+  func testCircularReferenceThreadSafety() {
+    container.register(.ObjectGraph) { ClientImp(service: try container.resolve()) as Client }
+    
+    let resolveClient: TMain = { _ in
+      let resolved = try! container.resolve() as Client
+      return UnsafeMutablePointer(Unmanaged.passUnretained(resolved as! ClientImp).toOpaque())
+    }
+    container.register(.ObjectGraph) { ServiceImp() as Service }
+      .resolveDependencies { container, service in
+        var clientPointer: UnsafeMutablePointer<Void> = nil
+        clientPointer = dispatch_sync(resolveClient)
+        let client = Unmanaged<ClientImp>.fromOpaque(COpaquePointer(clientPointer)).takeUnretainedValue()
+        service.client = client
+    }
+    
+    for _ in 0..<100 {
+      dispatch_async { _ in
+        let client = try! container.resolve() as Client
+        pthread_spin_lock(&lock)
+        resolvedClients.append(client as! ClientImp)
+        pthread_spin_unlock(&lock)
+        return nil
+      }
+    }
+    
+    sleep(2)
```

Comment:
Just to let al the threads finish. Of course joining each thread would be better (then dispatch_xxx should return thread's pid). Again it's only because of limitations of swift-core libs, in this case xctest does not provide async utils yet.